client/state/data-layer/wpcom-http/test/utils.js

```diff
@@ -236,4 +243,144 @@ describe( 'WPCOM HTTP Data Layer', () => {
+		const onSuccess = ( action, data ) => ( { type: 'SUCCESS', data } );
+		const onError = ( action, error ) => ( { type: 'FAILURE', error } );
+		const onProgress = ( action, progress ) => ( { type: 'PROGRESS', progress } );
+
+		const dispatcher = dispatchRequestEx( {
+			fetch,
+			onSuccess,
+			onError,
+			onProgress,
+		} );
+
+		const data = { count: 5 };
+		const error = { message: 'oh no!' };
+		const progress = { loaded: 45, total: 80 };
+
+		const fetchHttpAction = { type: 'REFILL' };
+		const successHttpAction = { type: 'REFILL', meta: { dataLayer: { data } } };
+		const failureHttpAction = { type: 'REFILL', meta: { dataLayer: { error } } };
+		const progressHttpAction = { type: 'REFILL', meta: { dataLayer: { progress } } };
+		const bothHttpAction = { type: 'REFILL', meta: { dataLayer: { data, error } } };
+
+		const fetchAction = { type: 'REQUEST' };
+		const successAction = { type: 'SUCCESS', data };
+		const failureAction = { type: 'FAILURE', error };
+		const progressAction = { type: 'PROGRESS', progress };
+
+		let dispatch;
+		let store;
+
+		beforeEach( () => {
+			dispatch = spy();
+			store = { dispatch };
+		} );
+
+		test( 'should initiate request if meta information missing', () => {
+			dispatcher( store, fetchHttpAction );
+			expect( dispatch ).to.have.been.calledWith( fetchAction );
+		} );
+
+		test( 'should call onSuccess if meta includes response data', () => {
+			dispatcher( store, successHttpAction );
+			expect( dispatch ).to.have.been.calledWith( successAction );
+		} );
+
+		test( 'should call onError if meta includes error data', () => {
+			dispatcher( store, failureHttpAction );
+			expect( dispatch ).to.have.been.calledWith( failureAction );
+		} );
+
+		test( 'should call onError if meta includes both response data and error data', () => {
+			dispatcher( store, bothHttpAction );
+			expect( dispatch ).to.have.been.calledWith( failureAction );
+		} );
+
+		test( 'should call onProgress if meta includes progress data', () => {
+			dispatcher( store, progressHttpAction );
+			expect( dispatch ).to.have.been.calledWith( progressAction );
+		} );
+
+		test( 'should not throw runtime error if onProgress is not specified', () => {
+			expect( () => {
+				dispatchRequestEx( { fetch, onSuccess, onError } )( store, progressHttpAction );
+			} ).to.not.throw( TypeError );
+		} );
+
+		test( 'should validate response data', () => {
+			const fromApi = makeParser( {
+				type: 'object',
+				properties: { count: { type: 'integer' } },
+			} );
+			dispatchRequestEx( { fetch, onSuccess, onError, fromApi } )( store, successHttpAction );
+			expect( dispatch ).to.have.been.calledWith( successAction );
+		} );
+
+		test( 'should fail-over on invalid response data', () => {
+			const fromApi = makeParser( {
+				type: 'object',
+				properties: { count: { type: 'string' } },
+			} );
+			dispatchRequestEx( { fetch, onSuccess, onError, fromApi } )( store, successHttpAction );
```

Comment:
some of these variables, particularly here, are making it hard for me to see what the test is doing - would there be any reason not to inline some of the actions/objects which are test-specific and contain the things the test is paying attention to?

for example, here I'm wondering about the `count` property - are we sending it in the test? having the object right here would help.

Reply:
If you have ideas to improve the tests, please don't hesitate and push a commit please. I definitely agree there's a room for improvement.